design/notifications.md

```diff
@@ -0,0 +1,360 @@
+    "hash":"e249ff3a681f177c"
+  }
+}
+```
+
+### Reference assignment
+Reference assignment events are triggered whenever a reference is assigned. In other words, whenever the 
+`PUT /trees/{referenceType}/{referenceName}` API in Nessie is called.
+
+#### Schema
+```yaml
+ReferenceAssigned:
+  type: object
+  properties:
+    type:
+      type: string
+      description: Always REFERENCE_ASSIGNED
+    eventTime:
+      type: string
+        format: date
+        description: The time the event is fired in ISO date format.
+    reference:
+      $ref: Reference
+      description: The reassigned reference.
+    assignedTo:
+      $ref: Reference 
+      description: The reference to which the reference got assigned to.
+```
+
+#### Example
+```json
+{
+  "type": "REFERENCE_ASSIGNED",
+  "eventTime": "2009-02-14T18:09:10Z",
+  "reference": {
+    "type": "BRANCH",
+    "name": "dev",
+    "hash": "e249ff3a681f177c"
+  },
+  "assignedTo": {
+    "type": "TAG",
+    "name": "release-10",
+    "hash": "e249ff3a681f177c"
+  }
+}
+```
+
+### Reference deletion
+Reference deletion events are triggered whenever a reference is deleted. In other words,  whenever the 
+`DELETE /trees/{referenceType}/{referenceName}` API in Nessie is called.
+
+#### Schema
+```yaml
+ReferenceDeleted:
+  type: object
+  properties:
+    type:
+      type: string
+      description: Always REFERENCE_DELETED
+    eventTime:
+      type: string
+      format: date
+      description: The time the event is fired in ISO date format.
+    referenceType:
+      type: string
+      description: The reference type. Either BRANCH or TAG.
+    referenceName:
+      type: string
+      description: The deleted reference name.
+```
+
+#### Example
+```json
+{
+  "type": "REFERENCE_DELETED",
+  "eventTime": "2009-02-14T18:09:10Z",
+  "referenceType": "BRANCH",
+  "referenceName": "main"
+}
+```
+
+## Webhooks/Notifications API
+A webhook is a method to augment the behavior of a web service with custom callbacks. These callbacks are usually a URL
+that a client registers with the webservice, in order for it to get invoked by the webservice whenever there's a
+relevant event.
+
+In Nessie's case, the contract between the webservice (caller) and the webhook URL (callee) is a simple HTTP request
+sent to the registered webhook. The contract is as follows:
+1. The client will register a webhook notification with Nessie by calling one of the APIs below.
+2. When an event happens in Nessie, it will call back using a `POST` request that:
+   1. Its content-type is `application/json`.
+   2. Its body contains the JSON payload of the event that happened. (See previous section for details on the expected schema).
+
+### Creating a new notification/webhook
+There are different event types a client might want to get notified about. In order to subscribe to a particular event 
+type, a client needs to make a `POST` request to one or more of the following endpoints.
+
+* `/notifications/commits`
```

Comment:
have you thought about the benefits/weaknesses of havnign a single endpoint and adding the type to the PUT request?

Reply:
do you mean like?

```
PUT /notifications

{
 eventType: "COMMIT"
 ...
}
```

OR merging the PUT and POST into a single operation and make the client choose the ID?

```
PUT /notifications/{client chosen ID}

{
 eventType: "COMMIT"
}
```

I like the idea for "futureproofing" for other types of events. I don't particularly like client generated IDs though.